3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DefaultAdyen3DS2Delegate.kt

```diff
@@ -0,0 +1,404 @@
+        when (action) {
+            is Threeds2FingerprintAction -> {
+                if (action.token.isNullOrEmpty()) {
+                    _exceptionFlow.tryEmit(ComponentException("Fingerprint token not found."))
+                    return
+                }
+                identifyShopper(
+                    activity = activity,
+                    encodedFingerprintToken = action.token.orEmpty(),
+                    submitFingerprintAutomatically = false,
+                    paymentData = paymentData,
+                )
+            }
+            is Threeds2ChallengeAction -> {
+                if (action.token.isNullOrEmpty()) {
+                    _exceptionFlow.tryEmit(ComponentException("Challenge token not found."))
+                    return
+                }
+                challengeShopper(activity, action.token.orEmpty())
+            }
+            is Threeds2Action -> {
+                if (action.token.isNullOrEmpty()) {
+                    _exceptionFlow.tryEmit(ComponentException("3DS2 token not found."))
+                    return
+                }
+                if (action.subtype == null) {
+                    _exceptionFlow.tryEmit(ComponentException("3DS2 Action subtype not found."))
+                    return
+                }
+                val subtype = Threeds2Action.SubType.parse(action.subtype.orEmpty())
+                // We need to keep authorizationToken in memory to access it later when the 3DS2 challenge is done
+                authorizationToken = action.authorisationToken
+                handleActionSubtype(activity, subtype, action.token.orEmpty(), paymentData)
+            }
+        }
+    }
+
+    private fun handleActionSubtype(
+        activity: Activity,
+        subtype: Threeds2Action.SubType,
+        token: String,
+        paymentData: String?,
+    ) {
+        when (subtype) {
+            Threeds2Action.SubType.FINGERPRINT -> identifyShopper(
+                activity = activity,
+                encodedFingerprintToken = token,
+                submitFingerprintAutomatically = true,
+                paymentData = paymentData,
+            )
+            Threeds2Action.SubType.CHALLENGE -> challengeShopper(activity, token)
+        }
+    }
+
+    @Suppress("LongMethod")
+    @VisibleForTesting
+    @Throws(ComponentException::class)
+    internal fun identifyShopper(
+        activity: Activity,
+        encodedFingerprintToken: String,
+        submitFingerprintAutomatically: Boolean,
+        paymentData: String?,
+    ) {
+        Logger.d(TAG, "identifyShopper - submitFingerprintAutomatically: $submitFingerprintAutomatically")
+        val decodedFingerprintToken = base64Encoder.decode(encodedFingerprintToken)
+
+        val fingerprintJson: JSONObject = try {
+            JSONObject(decodedFingerprintToken)
+        } catch (e: JSONException) {
+            throw ComponentException("JSON parsing of FingerprintToken failed", e)
```

Comment:
Should we replace all `throw` with flow emit in the delegate?

Reply:
I have been thinking about that as well, but I think it will make the code less readable. In most cases, the method that calls a method that throws an exception has to return in case of an exception. Meaning, the method that throws should return a boolean or something, so the calling method can return. It will some add extra if statements/overhead 🤔

Reply:
Good point